04-onion-routing.md

```diff
@@ -436,6 +458,94 @@ otherwise meets the amount criterion (eg. some other failure, or
+    - In the trampoline onion payload for the last trampoline node:
+      - MUST include a subset of the invoice's blinded paths in `recipient_blinded_paths`.
+      - MUST NOT include `outgoing_node_id`.
+      - SHOULD include the invoice features in `recipient_features`.
+  - Otherwise:
+    - MUST NOT use trampoline routing to pay that invoice.
+- MUST ensure that each hop in the `trampoline_onion_packet` supports `trampoline_routing`.
+- When paying a Bolt 11 invoice:
+  - MUST encrypt the `trampoline_onion_packet` with the same construction as `onion_packet`:
+    - MUST include `amt_to_forward` and `outgoing_cltv_value` for each hop.
+    - MUST use `outgoing_node_id` instead of `short_channel_id` to identify the next trampoline node.
+    - MUST include the invoice's `payment_secret` in the _last_ trampoline hop's payload.
+    - MAY add trailing filler data similar to what is done in the `onion_packet`.
+- When paying a Bolt 12 invoice that supports the `trampoline_routing` feature:
+  - For each node in the invoice's blinded path that the sender wants to use:
+    - MUST create a trampoline onion payload which:
+      - MUST include the `encrypted_recipient_data`.
+      - For the first node in the blinded route:
+        - MUST include the `current_path_key` provided in the invoice.
+      - For the final node:
+        - MUST include `amt_to_forward`, `outgoing_cltv_value` and `total_amount_msat`.
+      - MUST NOT include any other field.
+  - MAY prepend additional trampoline nodes where the trampoline onion payload:
+    - MUST include `amt_to_forward` and `outgoing_cltv_value`.
+    - MUST include `outgoing_node_id`.
+- MUST use a different `session_key` for the `trampoline_onion_packet` and the `onion_packet`.
+- MUST include the `trampoline_onion_packet` tlv in the _last_ hop's payload of the `onion_packet`.
+- If it sends a multi-part payment:
+  - MUST generate a random `payment_secret` to use in the outer onion.
+  - MUST NOT use the invoice's `payment_secret` in the outer onion.
+
+When processing a `trampoline_onion_packet`, a receiving node:
+
+- If it doesn't support `trampoline_routing`:
+  - MUST report a route failure to the origin node.
+- Otherwise, if it supports `trampoline_routing`:
+  - MUST process the `trampoline_onion_packet` as an `onion_packet`.
+  - MUST fail the HTLC if dictated by the requirements under [Failure Messages](#failure-messages).
+  - If it is not the final node:
+    - If the incoming payment is a multi-part payment:
+      - MUST wait to receive all the payment parts before forwarding.
+    - If `encrypted_recipient_data` is included:
+      - MUST use `current_path_key` from the trampoline onion or the outer onion to decrypt it.
+      - MUST validate its content as it would for the non-trampoline case.
+      - MUST include the next `current_path_key` in the `hop_payload` for the next trampoline node.
+    - MUST compute a route to the next trampoline node.
```

Comment:
Are there any thoughts on how to protect against DoS here? Also wondering what the most effective way would be for an attacker to overload trampoline nodes. I suppose a max length route sending the htlc in circles with each hop needing to search for a hard to find path would be a component of it.

It seems to be a channel jamming like problem because there's again no info on sender or receiver to filter out / limit trampoline activitiy.

Reply:
I am wondering what happens if a node doesn't filter for its own clients and just offers trampoline routing to anyone. Would it be advisable at all to enable this, or is this asking for trouble without having the heuristics in place too?

I wouldn't even know how to do these heuristics.